### PR TITLE
Replace `Any` in `optparse`

### PR DESCRIPTION
**Pull request by srittau**
The first commit replaces some `Any`s with the correct type and should probably be reviewed in isolation.

The second commit replaces all remaining `Any`s with `Incomplete` (or removes the annotation completely). This may be overzealous, as there might be legitimate `Any`s, but I opted for the "safer" side, as I don't have time to review them and they are all a bit suspect.

### CODE REVIEW

**Comment by srittau** [comment]
Regarding the primer output: The pip annotations need to be updated to include `None`. They code already handles the `None` case correctly.

**Comment by Avasam** [comment]
This PR works towards #9550

**Review by hamdanal — COMMENTED** [review]
I worked with optparse internals recently and I got somewhat familiar with it so I decided to leave you some suggestions that I thought could help your PR :)

File: stdlib/optparse.pyi [hunk 1]
```diff
@@ -140,28 +141,28 @@ class OptionContainer:
...
-    def _check_conflict(self, option: Any) -> None: ...
+    def __init__(self, option_class: type[Option], conflict_handler, description: str | None) -> None: ...
+    def _check_conflict(self, option) -> None: ...
     def _create_option_mappings(self) -> None: ...
     def _share_option_mappings(self, parser: OptionParser) -> None: ...
     @overload
     def add_option(self, opt: Option) -> Option: ...
     @overload
-    def add_option(self, *args: str | None, **kwargs: Any) -> Any: ...
+    def add_option(self, *args: str | None, **kwargs): ...
```
Comment (srittau) [inline stdlib/optparse.pyi:154]:
```suggestion
    def add_option(self, arg: str, /, *args: str | None, **kwargs) -> Option: ...
```

**Review by hamdanal — APPROVED** [review]
LGTM, thank you.

Both mypy primer hits in pip are true positives. Their functions accept `None` but are annotated as accepting only `str` (see [here](https://github.com/pypa/pip/blob/36823099a9cdd83261fdbc8c1d2a24fa2eea72ca/src/pip/_internal/cli/parser.py#L70-L93)); I guess they were just following the old optparse annotations.

**Pull request merged**